Sources/SwiftFormatPrettyPrint/SequenceExprFolding.swift

```diff
@@ -0,0 +1,490 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2019 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SwiftSyntax
+
+extension SequenceExprSyntax {
+
+  /// Returns a transformed version of the `SequenceExprSyntax` such that its
+  /// shape is tree-like based on the precedence and associativity of the
+  /// operators used therein.
+  ///
+  /// Once folded, the returned expression will have one of three possible
+  /// shapes:
+  ///
+  /// - A subtype of `ExprSyntax` other than `SequenceExprSyntax`
+  /// - A `SequenceExprSyntax` with exactly two elements, where the first is an
+  ///   operand (a subtype of `ExprSyntax`) and the second is a cast expression
+  ///   (either `AsExprSyntax` or `IsExprSyntax`)
+  /// - A `SequenceExprSyntax` with exactly three elements, where the first and
+  ///   third elements are operands (subtypes of `ExprSyntax`) and the second is
+  ///   a `BinaryOperatorExprSyntax`
```

Comment:
As part of moving this to SwiftSyntax I'd recommend that we add new syntax nodes (like `BinaryExprSyntax`) that completely replace `SequenceExprSyntax` from the tree. `SequenceExprSyntax` should not be present in a folded tree then (unless there was an unknown operator).

Reply:
👍 to this idea; for now, I've added a TODO to track doing this so that we can move forward with other changes that depend on this.